log4j-core/src/main/java/org/apache/logging/log4j/core/jackson/Initializers.java

```diff
@@ -62,33 +65,43 @@ void setupModule(
                             : ThrowableProxyWithoutStacktraceMixIn.class);
         }
     }
+    /**
+     * Used to set up {@link SetupContext} from different {@link SimpleModule}s.
+     * <p>
+     *     Serializes the context map as list of objects.
+     * </p>
+     */
+    static class SetupContextAsEntryListInitializer extends AbstractInitializer {
+
+        @Override
+        void setupModule(
+                final SetupContext context, final boolean includeStacktrace, final boolean stacktraceAsString) {
+            super.setupModule(context, includeStacktrace, stacktraceAsString);
+            // Prevents reflective JPMS access
+            // https://github.com/apache/logging-log4j2/issues/2814
+            context.addSerializers(new SimpleSerializers(singletonList(new ContextDataAsEntryListSerializer())));
+            context.addDeserializers(
+                    new SimpleDeserializers(singletonMap(StringMap.class, new ContextDataAsEntryListDeserializer())));
```

Comment:
Why don't we need the same for, say, `Log4jStackTraceElementDeserializer`?

Reply:
That one is a public class with a public constructor.
I am not very eager to expose additional classes to the public API.

**BTW**: I am pretty sure we don't need to deserialize JSON in Log4j Core. We do it in unit tests and `log4j-server` used it, but the rest should be happy with serialization alone.

Unfortunately I don't think it is easy to remove the deserialization feature.

Reply:
> That one is a public class with a public constructor. I am not very eager to expose additional classes to the public API.

Right. That is a very subtle design detail. In particular, now we deviate from the existing practice, i.e., making all stuff public. Would you mind (preferably) in detail explaining this in the comments, please? I can imagine it can be very obvious to you, but I can assure you certainly not to the rest of the maintainers.

Reply:
Fixed in https://github.com/apache/logging-log4j2/pull/2815/commits/0b8c1955765d47fa978012b4c1b3b101169e88fe

Note that I don't have any idea what the best practices for encapsulating classes related to Jackson are.
I am merely trying to maintain the spirit of the original authors.

Reply:
Thanks. I think original authors would have just marked those (de)serializers as public and be done with it. :shrug: